Error - Unsupported platform: freebsd

### Instruction on how to enable debug and html trace

(Applicable to Docker only?)


### Environment

**FlareSolverr Version:** v1.2.3

**Docker:** No

**OS:** FreeBSD 11.4 (Jail under TrueNAS Core 12.0-U2.1 (FreeBSD 12.2))

**Last Working FlareSolverr Version:** N/A

**Are you using a proxy or VPN?:** Yes - VPN (at router level)

**Using Captcha Solver:** No

**If using captcha solver, which one:** N/A

**Mono Version:** 6.8.0.105
**Jackett Version:** v0.17.617
**Node Version:** v15.5.1
**npm Version:** 6.14.8


### Description

Installed node & npm through native pkg commands. 
`npm install` fails. Missing fsevents. Might be other major issues here? Forced installation with `-f`.
`npm run build`, then, `npm start`.
Successfully connects from jackett, however, all related indexer tests fail due to Unsupported platform Error.


### Logged Error Messages

root@:/usr/local/share/flaresolverr # npm start

> flaresolverr@1.2.3 start /usr/local/share/flaresolverr
> node ./dist/index.js

2021-03-05T03:00:19.003Z INFO REQ-0 FlareSolverr v1.2.3 listening on http://0.0.0.0:8191
2021-03-05T03:00:28.295Z INFO REQ-1 Incoming request: POST /v1
2021-03-05T03:00:28.296Z INFO REQ-1 Params: {"maxTimeout":60000,"cmd":"request.get","url":"https://www.7torrents.cc/search?query=+&sort=created","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36"}
Error: Unsupported platform: freebsd
    at Object.assert (/usr/local/share/flaresolverr/node_modules/puppeteer/lib/helper.js:29:15)
    at BrowserFetcher.setPlatform (/usr/local/share/flaresolverr/node_modules/puppeteer/lib/BrowserFetcher.js:119:22)
    at new BrowserFetcher (/usr/local/share/flaresolverr/node_modules/puppeteer/lib/BrowserFetcher.js:103:14)
    at resolveExecutablePath (/usr/local/share/flaresolverr/node_modules/puppeteer/lib/Launcher.js:488:28)
[...]

2021-03-05T03:20:02.126Z INFO REQ-47 Incoming request: POST /v1
2021-03-05T03:20:02.127Z INFO REQ-47 Params: {"maxTimeout":60000,"cmd":"request.get","url":"https://apibay.org/precompiled/data_top100_recent.json","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36"}
Error: Unsupported platform: freebsd
    at Object.assert (/usr/local/share/flaresolverr/node_modules/puppeteer/lib/helper.js:29:15)
    at BrowserFetcher.setPlatform (/usr/local/share/flaresolverr/node_modules/puppeteer/lib/BrowserFetcher.js:119:22)
    at new BrowserFetcher (/usr/local/share/flaresolverr/node_modules/puppeteer/lib/BrowserFetcher.js:103:14)
    at resolveExecutablePath (/usr/local/share/flaresolverr/node_modules/puppeteer/lib/Launcher.js:488:28)
    at ChromeLauncher.launch (/usr/local/share/flaresolverr/node_modules/puppeteer/lib/Launcher.js:57:53)
    at async PuppeteerExtra.launch (/usr/local/share/flaresolverr/node_modules/puppeteer-extra/dist/index.cjs.js:129:25)
    at async Object.create (/usr/local/share/flaresolverr/dist/session.js:49:27)
    at async browserRequest (/usr/local/share/flaresolverr/dist/routes.js:138:11)
    at async request.get (/usr/local/share/flaresolverr/dist/routes.js:189:9)
    at async Object.Router [as default] (/usr/local/share/flaresolverr/dist/routes.js:202:16)
2021-03-05T03:20:02.127Z ERROR REQ-47 Unsupported platform: freebsd

Tried to install this into 12.2-RELEASE-p4 (iocage jail) on TrueNAS-12.0-U7 (host).
pkg install chrome git
cloned repo
followed the env settings and ran npm install, npm run build and then npm start, got errors and finally i even tried an install with -f

This is what i got.
```
> flaresolverr@2.1.0 start /root/FlareSolverr
> node ./dist/server.js

2021-12-28T19:12:57+01:00 INFO FlareSolverr v2.1.0
2021-12-28T19:12:57+01:00 INFO Testing web browser installation...
Assertion failed: Unsupported platform:
2021-12-28T19:12:57+01:00 ERROR Unsupported platform:
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! flaresolverr@2.1.0 start: `node ./dist/server.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the flaresolverr@2.1.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-12-28T18_12_57_410Z-debug.log
```

log has this in it.
```
~/FlareSolverr # cat /root/.npm/_logs/2021-12-28T18_12_57_410Z-debug.log
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/bin/node', '/usr/local/bin/npm', 'start' ]
2 info using npm@6.14.8
3 info using node@v14.18.1
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle flaresolverr@2.1.0~prestart: flaresolverr@2.1.0
6 info lifecycle flaresolverr@2.1.0~start: flaresolverr@2.1.0
7 verbose lifecycle flaresolverr@2.1.0~start: unsafe-perm in lifecycle true
8 verbose lifecycle flaresolverr@2.1.0~start: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/root/FlareSolverr/node_modules/.bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin:/root/bin
9 verbose lifecycle flaresolverr@2.1.0~start: CWD: /root/FlareSolverr
10 silly lifecycle flaresolverr@2.1.0~start: Args: [ '-c', 'node ./dist/server.js' ]
11 silly lifecycle flaresolverr@2.1.0~start: Returned: code: 1  signal: null
12 info lifecycle flaresolverr@2.1.0~start: Failed to exec start script
13 verbose stack Error: flaresolverr@2.1.0 start: `node ./dist/server.js`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:400:28)
13 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:400:28)
13 verbose stack     at maybeClose (internal/child_process.js:1058:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:293:5)
14 verbose pkgid flaresolverr@2.1.0
15 verbose cwd /root/FlareSolverr
16 verbose FreeBSD 12.2-RELEASE-p11
17 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm" "start"
18 verbose node v14.18.1
19 verbose npm  v6.14.8
20 error code ELIFECYCLE
21 error errno 1
22 error flaresolverr@2.1.0 start: `node ./dist/server.js`
22 error Exit status 1
23 error Failed at the flaresolverr@2.1.0 start script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

Nicely done @Zamana , altho i get this error on a freshly made iocage jail.

```
> flaresolverr@2.1.0 start /usr/local/share/FlareSolverr
> node ./dist/server.js

internal/modules/cjs/loader.js:905
  throw err;
  ^

Error: Cannot find module '/usr/local/share/FlareSolverr/dist/server.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
    at Function.Module._load (internal/modules/cjs/loader.js:746:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! flaresolverr@2.1.0 start: `node ./dist/server.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the flaresolverr@2.1.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-12-30T12_19_02_178Z-debug.log
```

--- a/README.md
+++ b/README.md
@@ -83,6 +83,13 @@ This is the recommended way for Windows users.
 * Run `pip install -r requirements.txt` command to install FlareSolverr dependencies.
 * Run `python src/flaresolverr.py` command to start FlareSolverr.
 
+### From source code (FreeBSD/TrueNAS CORE)
+
+* Run `pkg install chromium python39 py39-pip xorg-vfbserver` command to install the required dependencies.
+* Clone this repository and open a shell in that path.
+* Run `python3.9 -m pip install -r requirements.txt` command to install FlareSolverr dependencies.
+* Run `python3.9 src/flaresolverr.py` command to start FlareSolverr.
+
 ### Systemd service
 
 We provide an example Systemd unit file `flaresolverr.service` as reference. You have to modify the file to suit your needs: paths, user and environment variables.

--- a/src/undetected_chromedriver/__init__.py
+++ b/src/undetected_chromedriver/__init__.py
@@ -451,8 +451,10 @@ class Chrome(selenium.webdriver.chrome.webdriver.WebDriver):
                 options.binary_location, *options.arguments
             )
         else:
-            startupinfo = subprocess.STARTUPINFO()
+            startupinfo = None
             if os.name == 'nt' and windows_headless:
+                # STARTUPINFO() is Windows only
+                startupinfo = subprocess.STARTUPINFO()
                 startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
             browser = subprocess.Popen(
                 [options.binary_location, *options.arguments],

--- a/src/undetected_chromedriver/patcher.py
+++ b/src/undetected_chromedriver/patcher.py
@@ -21,7 +21,7 @@ from multiprocessing import Lock
 
 logger = logging.getLogger(__name__)
 
-IS_POSIX = sys.platform.startswith(("darwin", "cygwin", "linux", "linux2"))
+IS_POSIX = sys.platform.startswith(("darwin", "cygwin", "linux", "linux2", "freebsd"))
 
 
 class Patcher(object):
@@ -80,9 +80,14 @@ class Patcher(object):
             os.makedirs(self.data_path, exist_ok=True)
 
         if not executable_path:
-            self.executable_path = os.path.join(
-                self.data_path, "_".join([prefix, self.exe_name])
-            )
+            if sys.platform.startswith("freebsd"):
+                self.executable_path = os.path.join(
+                    self.data_path, self.exe_name
+                )
+            else:
+                self.executable_path = os.path.join(
+                    self.data_path, "_".join([prefix, self.exe_name])
+                )
 
         if not IS_POSIX:
             if executable_path:
@@ -127,6 +132,9 @@ class Patcher(object):
             else:
                 self.platform_name = "mac-x64"
             self.exe_name %= ""
+        if self.platform.startswith("freebsd"):
+            self.platform_name = "freebsd"
+            self.exe_name %= ""
 
     def auto(self, executable_path=None, force=False, version_main=None, _=None):
         """
@@ -166,26 +174,56 @@ class Patcher(object):
         if force is True:
             self.force = force
 
-        try:
-            os.unlink(self.executable_path)
-        except PermissionError:
-            if self.force:
-                self.force_kill_instances(self.executable_path)
-                return self.auto(force=not self.force)
-            try:
-                if self.is_binary_patched():
-                    # assumes already running AND patched
-                    return True
-            except PermissionError:
-                pass
-            # return False
-        except FileNotFoundError:
-            pass
 
-        release = self.fetch_release_number()
-        self.version_main = release.version[0]
-        self.version_full = release
-        self.unzip_package(self.fetch_package())
+        if self.platform_name == "freebsd":
+            chromedriver_path = shutil.which("chromedriver")
+
+            if not os.path.isfile(chromedriver_path) or not os.access(chromedriver_path, os.X_OK):
+                logging.error("Chromedriver not installed!")
+                return
+
+            version_path = os.path.join(os.path.dirname(self.executable_path), "version.txt")
+
+            process = os.popen(f'"{chromedriver_path}" --version')
+            chromedriver_version = process.read().split(' ')[1].split(' ')[0]
+            process.close()
+
+            current_version = None
+            if os.path.isfile(version_path) or os.access(version_path, os.X_OK):
+                with open(version_path, 'r') as f:
+                    current_version = f.read()
+
+            if current_version != chromedriver_version:
+                logging.info("Copying chromedriver executable...")
+                shutil.copy(chromedriver_path, self.executable_path)
+                os.chmod(self.executable_path, 0o755)
+
+                with open(version_path, 'w') as f:
+                    f.write(chromedriver_version)
+
+                logging.info("Chromedriver executable copied!")
+        else:
+            try:
+                os.unlink(self.executable_path)
+            except PermissionError:
+                if self.force:
+                    self.force_kill_instances(self.executable_path)
+                    return self.auto(force=not self.force)
+                try:
+                    if self.is_binary_patched():
+                        # assumes already running AND patched
+                        return True
+                except PermissionError:
+                    pass
+                # return False
+            except FileNotFoundError:
+                pass
+
+            release = self.fetch_release_number()
+            self.version_main = release.version[0]
+            self.version_full = release
+            self.unzip_package(self.fetch_package())
+
         return self.patch()
 
     def driver_binary_in_use(self, path: str = None) -> bool:

--- a/src/utils.py
+++ b/src/utils.py
@@ -5,6 +5,7 @@ import re
 import shutil
 import urllib.parse
 import tempfile
+import sys
 
 from selenium.webdriver.chrome.webdriver import WebDriver
 import undetected_chromedriver as uc
@@ -157,7 +158,7 @@ def get_webdriver(proxy: dict = None) -> WebDriver:
     # we launch the browser in head-full mode with the window hidden
     windows_headless = False
     if get_config_headless():
-        if os.name == 'nt':
+        if os.name == 'nt' or sys.platform.startswith('freebsd'):
             windows_headless = True
         else:
             start_xvfb_display()
@@ -180,9 +181,12 @@ def get_webdriver(proxy: dict = None) -> WebDriver:
 
     # downloads and patches the chromedriver
     # if we don't set driver_executable_path it downloads, patches, and deletes the driver each time
-    driver = uc.Chrome(options=options, browser_executable_path=browser_executable_path,
-                       driver_executable_path=driver_exe_path, version_main=version_main,
-                       windows_headless=windows_headless, headless=windows_headless)
+    try:
+        driver = uc.Chrome(options=options, browser_executable_path=browser_executable_path,
+                           driver_executable_path=driver_exe_path, version_main=version_main,
+                           windows_headless=windows_headless, headless=windows_headless)
+    except Exception as e:
+        logging.error("Error starting Chrome: %s" % e)
 
     # save the patched driver to avoid re-downloads
     if driver_exe_path is None:

--- a/src/utils.py
+++ b/src/utils.py
@@ -158,7 +158,7 @@ def get_webdriver(proxy: dict = None) -> WebDriver:
     # we launch the browser in head-full mode with the window hidden
     windows_headless = False
     if get_config_headless():
-        if os.name == 'nt' or sys.platform.startswith('freebsd'):
+        if os.name == 'nt':
             windows_headless = True
         else:
             start_xvfb_display()

